extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java

```diff
@@ -31,15 +32,23 @@ public GrpcRequestContextGrpcInterceptor() {
         // This interceptor is called first, so, we should be on the event loop.
         Context capturedVertxContext = Vertx.currentContext();
         if (capturedVertxContext != null) {
-            GrpcRequestContextHolder contextHolder = GrpcRequestContextHolder.initialize(capturedVertxContext);
+            InjectableContext.ContextState state;
+            if (!reqContext.isActive()) {
+                reqContext.activate();
+                state = reqContext.getState();
+            } else {
+                state = null;
+            }
+
             ServerCall.Listener<ReqT> delegate = next
                     .startCall(new ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT>(call) {
 
                         @Override
                         public void close(Status status, Metadata trailers) {
                             super.close(status, trailers);
-                            if (contextHolder.state != null) {
-                                reqContext.destroy(contextHolder.state);
+                            if (state != null) {
+                                reqContext.deactivate();
+                                capturedVertxContext.runOnContext(event -> reqContext.destroy(state));
```

Comment:
I think the proper sequence is `destroy` then `deactivate`. Also, why deactivate synchronously and destroy asynchronously?

Reply:
hmm... I took a deeper look at the code and it seems `destroy` uses the context from a thread local. And `deactivate` cleans this threadlocal... 
Given that `destroy` cleans up the context, I think it's enough to call it. 
`capturedVertxContext.runOnContext` was to make sure it is executed with the same context as the creation of the request context

Reply:
I removed `deactivate`

Reply:
I'd really like to have that vetted by @mkouba, because my understanding is that if you don't `deactivate`, you leave thread-local state behind.

Reply:
`deactivate` just removes the state from the current thread, it doesn't fire destroy for beans.

If the map is cleaned, why can't it be reused?

Reply:
Because it may be captured somewhere else for context propagation? (Not exactly sure how _that_ works.)

Reply:
Ah and good point re destroying beans. In that case, IMHO both `destroy` and `deactivate` must be called.

Reply:
Both `destroy()` and `deactivate()` should be always called in the end. There's even `ManagedContext.terminate()` that should be used in simple cases. The reason why there are two separate methods is that in some situations you need to _deactivate_ the context (i.e. remove the thread local) but not _destroy_ the bean instances.

Reply:
Thanks guys, I'll restore the original order to have first `deactivate` and then `destroy` then

Reply:
I think it's correct now and I'll let myself resolve this comment :)